### PR TITLE
Allows editing of start/end tag simultaneously

### LINKED ISSUE #130: Auto Rename Tag
I just realized, with the help of [@angelozerr](https://github.com/angelozerr) at [lsp4xml/#362](https://github.com/angelozerr/lsp4xml/issues/362), that synced renaming and deletion of two tags, that belong together, may best be done via some simple text manipulation on the client side. I now have installed a little extension into VSCode, that does just that (minus paired delete, sadly), but it may well fit also into your project, so to keep things together.

### PR DESCRIPTION
**Pull request by NikolasKomonen**
Under the preference xml.autoSelectingMatchingTags

Fixes #130

Signed-off-by: Nikolas Komonen <nikolaskomonen@gmail.com>

### CODE REVIEW

File: src/matchingTag.ts [hunk 1]
```diff
@@ -0,0 +1,147 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (angelozerr) [inline src/matchingTag.ts]: @NikolasKomonen it seems that you are using an old code (which I think that there are some bugs). IMHO you should use https://github.com/microsoft/vscode/blob/master/extensions/html-language-features/client/src/mirrorCursor.ts

**Comment by fbricon** [comment]
Can't get it to work. Output filled with errors like:
```
rejected promise not handled within 1 second: Error: Unsupported request method: xml/matchingTagPosition
extensionHostProcess.js:820
stack trace: Error: Unsupported request method: xml/matchingTagPosition
	at handleResponse (/Users/fbricon/Dev/projects/vscode-xml/dist/extension.js:6179:48)
	at processMessageQueue (/Users/fbricon/Dev/projects/vscode-xml/dist/extension.js:6006:17)
	at Immediate.<anonymous> (/Users/fbricon/Dev/projects/vscode-xml/dist/extension.js:5990:13)
	at processImmediate (internal/timers.js:439:21)
```

**Comment by fbricon** [comment]
When typing a space at the end of a tag, to add an attribute for instance, there's often an error logged because of cancelled request. Logs fill up pretty quickly so this should be handled more gracefully:
Dec-10-2019 16-16-23

**Comment by NikolasKomonen** [comment]
> When typing a space at the end of a tag, to add an attribute for instance, there's often an error logged because of cancelled request. Logs fill up pretty quickly so this should be handled more gracefully:

@fbricon I can't recreate this, but maybe it is using the bugged version of the code. I pushed the new one here, just pull the client again and hopefully this disappears.

**Comment by fbricon** [comment]
Works beautifully now, can't reproduce the issue from earlier. 

That's a great feature, nice job @NikolasKomonen!

**Comment by fbricon** [comment]
Spoke too soon, I can still reproduce. add space at the end of a tag, delete, rinse, repeat until the exception shows up

**Comment by fbricon** [comment]
definitely reminds me of #149. We prolly need to make the server side more robust wrt cancelled commands

**Comment by fbricon** [comment]
I think the key here is "rejected promise not handled within 1 second". Server is too slow to respond, then client cancels the request.

**Comment by fbricon** [comment]
See https://stackoverflow.com/a/50124430/753170 for a potential solution

**Comment by NikolasKomonen** [comment]
> See https://stackoverflow.com/a/50124430/753170 for a potential solution

Yup, this worked for me.

See: https://github.com/redhat-developer/vscode-xml/pull/204/files#diff-2a8a3831675ce7f9a5318a33b0485d7dR86

**Comment by fbricon** [comment]
No this doesn't work for me

**Comment by fbricon** [comment]
Now that've been playing with it, I think we need a new command to toggle the preference, so that, when bound to a keystroke, you can enter/exit quick tag edit mode

Also, the readme needs to be updated with the preference key (and documenting that new toggle command would help too)

**Comment by NikolasKomonen** [comment]
>  I think we need a new command to toggle the preference

Right now you can press Esc to stop the edit mode. Is that good enough?

Also, I have the README updated here in the PR here: https://github.com/redhat-developer/vscode-xml/pull/206/files#diff-04c6e90faac2675aa89e2176d2eec7d8R82

**Comment by fbricon** [comment]
Ctrl+Shift+F2 on linux/windows, Cmd+Shift+F2 on mac, to toggle matching tag editing

File: package.json [hunk 2]
```diff
@@ -192,6 +192,12 @@
           "description": "Enable/disable autoclosing of XML tags. \n\nIMPORTANT: Turn off editor.autoClosingTags for this to work",
           "scope": "window"
         },
+        "xml.autoSelectingMatchingTags": {
```
Comment (fbricon) [inline package.json]: xml.matchingTagEditing

File: src/mirrorCursor.ts [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+							const newCursor = new Selection(
+								matchingTagPosition.line,
+								matchingTagPosition.character,
+								matchingTagPosition.line,
+								matchingTagPosition.character
+							);
+							window.activeTextEditor.selections = [...window.activeTextEditor.selections, newCursor];
+						}
+					}
+				}).then(err => { return; }); // mutes "rejected promise not handled within 1 second"
```
Comment (fbricon) [inline src/mirrorCursor.ts]:
```typescript
.then(undefined,  err => {
	const msg = err.message ;
	// mutes "rejected promise not handled within 1 second"
	if (msg && !msg.endsWith('has been cancelled')){
		console.log(err);
	}
	return; 
});
```

**Comment by NikolasKomonen** [comment]
@fbricon updated everything. The console error should be gone, cmd+shift+f2 should toggle the preference on your machine and a message should appear.

**Comment by fbricon** [comment]
please update the readme

**Comment by NikolasKomonen** [comment]
> please update the readme

@fbricon the readme is updated in this PR: https://github.com/redhat-developer/vscode-xml/pull/206

**Comment by fbricon** [comment]
on configuration change, toggling the new pref should add/remove the mirror cursor immediately

File: package.json [hunk 4]
```diff
@@ -192,6 +192,12 @@
           "description": "Enable/disable autoclosing of XML tags. \n\nIMPORTANT: Turn off editor.autoClosingTags for this to work",
           "scope": "window"
         },
+        "xml.matchingTagEditing": {
```
Comment (angelozerr) [inline package.json]: what about using the same name than vscode `mirrorCursorOnMatchingTag` ?

**Comment by xorye** [comment]
> on configuration change, toggling the new pref should add/remove the mirror cursor immediately

I implemented that here: https://github.com/xorye/vscode-xml/tree/matchingTagClient

**Comment by fbricon** [comment]
superseded by #207

**Pull request closed without merging**